### PR TITLE
[FLINK-26371][hive] Support variable substitution for sql statement while using Hive dialect

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To support variable substitution for sql statement while using Hive dialect.

## Brief change log
Substitute the variable with actual value in sql statement while using Hive dialect before really parse it.

## Verifying this change
Additional test in [HiveDialectITCase.java](https://github.com/apache/flink/compare/master...luoyuxia:FLINK-26371?expand=1#diff-748a9fa63f15b3bd3ea5426a5b42179cadee77ecf3359c83f846aedb27d96871)#testStatementVariableSubstitution

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive):  no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?  N/A

### CODE REVIEW

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 1]
```diff
@@ -807,6 +807,29 @@ public void testShowPartitions() throws Exception {
...
+    public void testStatementVariableSubstitution() {
+        // test system variable for substitution
+        System.setProperty("k1", "v1");
+        List<Row> result =
+                CollectionUtil.iteratorToList(
+                        tableEnv.executeSql("select '${system:k1}'").collect());
+        assertEquals("[+I[v1]]", result.toString());
+
+        // test env variable for substitution
+        String classPath = System.getenv("CLASSPATH");
```
Comment (Tartarus0zm) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:820]: env `CLASSPATH ` maybe not exist.
We should test for an environment variable that exists and an environment variable that does not exist

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java [hunk 2]
```diff
@@ -807,6 +807,29 @@ public void testShowPartitions() throws Exception {
...
+        result =
+                CollectionUtil.iteratorToList(
+                        tableEnv.executeSql("select '${env:CLASSPATH}'").collect());
+        assertEquals(String.format("[+I[%s]]", classPath), result.toString());
+
+        // test hive conf variable for substitution
+        result =
+                CollectionUtil.iteratorToList(
+                        tableEnv.executeSql("select '${hiveconf:common-key}'").collect());
```
Comment (Tartarus0zm) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:829]: How do users dynamically set hiveconf variables in the flink table api? `common-key` is a variable that exists in the test hive-site.xml

Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectITCase.java:829]: Currently, we have no way to dynamically set hiveconf variables in the flink table api.  AFAK, hive only supports set hiveconf variables in cli or init a hive session with hiveconf variables.

**Comment by luoyuxia** [comment]
@flinkbot run azure

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**